src/UserManager.php

```diff
@@ -0,0 +1,353 @@
+          throw new \RuntimeException(sprintf('Cannot load user by field %s', $userIdField));
+        }
+        if (NULL !== $user) {
+          $this->logger->info($this->t('Retaining user @name.', ['@name' => $user->getAccountName()]));
+          unset($userIds[$user->id()]);
+        }
+      }
+      $this->setUserIds($userIds);
+    }
+  }
+
+  /**
+   * Delete users.
+   */
+  public function deleteUsers() {
+    $method = $this->moduleConfig->get('user_cancel_method');
+    $deletedUserIds = [];
+    $userIds = $this->getUserIds();
+    foreach ($userIds as $userId) {
+      user_cancel([], $userId, $method);
+      $deletedUserIds[] = $userId;
+    }
+
+    $this->logger->info($this->formatPlural(
+      count($deletedUserIds),
+      'One user to be deleted',
+      '@count users to be deleted'
+    ));
+
+    if (!($this->options['dry-run'] ?? FALSE)) {
+      if (!empty($deletedUserIds)) {
+        $this->logger->info($this->formatPlural(
+          count($deletedUserIds),
+          'Deleting one user',
+        'Deleting @count users'
+        ));
+        $this->requestStack->getCurrentRequest()
+          // batch_process needs a route in the request (!)
+          ->attributes->set(RouteObjectInterface::ROUTE_OBJECT, new Route('<none>'));
+
+        // Process the batch created by deleteUser.
+        $batch =& batch_get();
+        $batch['progressive'] = FALSE;
+        $batch['source_url'] = 'cron';
+
+        batch_process();
+      }
+    }
+  }
+
+  /**
+   * Load user by properties.
+   *
+   * @param array $properties
+   *   The properties.
+   *
+   * @return \Drupal\user\UserInterface|null
+   *   The user if any.
+   */
+  private function loadUserByProperties(array $properties = []): ?UserInterface {
+    /** @var \Drupal\user\UserInterface[] $users */
+    $users = $this->userStorage->loadByProperties($properties);
+
+    return reset($users) ?: NULL;
+  }
+
+  /**
+   * Get module user ids.
+   *
+   * @param string $module
+   *   The module.
+   *
+   * @return int[]|null
+   *   The user ids.
+   */
+  private function getModuleUsersIds(string $module): ?array {
+    if (!$this->moduleHandler->moduleExists($module)) {
+      return NULL;
+    }
+    switch ($module) {
+      case 'openid_connect':
+        return $this->database
+          ->select('users_data')
+          ->fields('users_data', ['uid'])
+          ->condition('users_data.module', 'openid_connect')
+          ->condition('users_data.name', 'oidc_name')
+          ->execute()
+          ->fetchCol();
+
+      case 'samlauth':
+        return $this->database
+          ->select('authmap')
+          ->fields('authmap', ['uid'])
+          ->condition('authmap.provider', 'samlauth')
+          ->execute()
+          ->fetchCol();
+
+      default:
+        throw new \RuntimeException(sprintf('Unknown module: %s', $module));
+    }
+  }
+
+  /**
+   * Set user ids.
+   */
+  private function setUserIds(array $userIds) {
```

Comment:
This is not a "setter". Would be better name as `cacheUserIdsForDeletion` or similar.